### ISSUE TITLE
ENH Improve `DatetimeEncoder`

### ISSUE THREAD

**Issue by Vincent-Maladiere** [skrub-data/skrub#768]
Recently, several bugs have been discovered with `DatetimeEncoder` (https://github.com/skrub-data/skrub/issues/741, https://github.com/skrub-data/skrub/issues/744, https://github.com/skrub-data/skrub/issues/745, https://github.com/skrub-data/skrub/issues/746), so this estimator is a good candidate for thorough refactoring.

https://github.com/skrub-data/skrub/pull/743 raised the following points to be addressed:
- Parse date column only once (when `DatetimeEncoder` is used within the `TableVectorizer`)
- Exposes the parsing result with `TableVectorizer` to unify the logic
- Avoid removing constant columns, since the shape of the dataframe varies (having it as a keyword argument)
	- Should we add the `VarianceThreshold` estimator to the example to remove constant features?

**Comment by jeromedockes** [comment]
One other thing we may want to fix is that in `pd.DateTimeIndex` (and thus `skrub.DateTimeEncoder`) day first vs month first is inferred separately for each row, which seems suboptimal because most tables probably don't mix the 2:
```python
>>> pd.DatetimeIndex(["28/01/2023", "04/07/2024"])
DatetimeIndex(['2023-01-28', '2024-04-07'], dtype='datetime64[ns]', freq=None)
>>> 
```

**Comment by jeromedockes** [comment]
Also if we decide to look at the whole column (or at least a chunk of it) as the TableVectorizer does ATM to determine the format, we would probably want to store that state during fit so that the same format is used for transform. this could lead to an organization like:
- datetimeencoder fit: find datetime columns and their format, expose the list of found columns in a public attribute
- tablevectorizer: use this list to decide which transformer to apply to which columns
- datetimeencoder transform: extract features from datetimecolumns

or even have separate estimators for parsing string columns that contain dates and for extracting features from datetime columns

**Comment by Vincent-Maladiere** [comment]
> One other thing we may want to fix is that in pd.DateTimeIndex (and thus skrub.DateTimeEncoder) day first vs month first is inferred separately for each row, which seems suboptimal because most tables probably don't mix the 2:
>  ```python
>  pd.DatetimeIndex(["28/01/2023", "04/07/2024"])
>  ```

Yes indeed, good catch. `to_datetime()` behaves as expected (with a warning `dayfirst` in this instance)
```python
>>> pd.to_datetime(["28/01/2023", "04/07/2024"])
DatetimeIndex(['2023-01-28', '2024-07-04'], dtype='datetime64[ns]', freq=None)
```
